esphome/components/ltr_als_ps/ltr_als_ps.cpp

```diff
@@ -0,0 +1,520 @@
+    return DataAvail::BAD_DATA;
+  }
+  ESP_LOGV(TAG, "Data ready, reported gain is %.0f", get_gain_coeff(als_status.gain));
+  if (data.gain != als_status.gain) {
+    ESP_LOGW(TAG, "Actual gain differs from requested (%.0f)", get_gain_coeff(data.gain));
+    return DataAvail::BAD_DATA;
+  }
+  return DataAvail::DATA_OK;
+}
+
+void LTRAlsPsComponent::read_sensor_data_(AlsReadings &data) {
+  data.ch1 = 0;
+  data.ch0 = 0;
+  uint8_t ch1_0 = this->reg((uint8_t) CommandRegisters::ALS_DATA_CH1_0).get();
+  uint8_t ch1_1 = this->reg((uint8_t) CommandRegisters::ALS_DATA_CH1_1).get();
+  uint8_t ch0_0 = this->reg((uint8_t) CommandRegisters::ALS_DATA_CH0_0).get();
+  uint8_t ch0_1 = this->reg((uint8_t) CommandRegisters::ALS_DATA_CH0_1).get();
+  data.ch1 = encode_uint16(ch1_1, ch1_0);
+  data.ch0 = encode_uint16(ch0_1, ch0_0);
+
+  ESP_LOGV(TAG, "Got sensor data: CH1 = %d, CH0 = %d", data.ch1, data.ch0);
+}
+
+bool LTRAlsPsComponent::are_adjustments_required_(AlsReadings &data) {
+  if (!this->automatic_mode_enabled_)
+    return false;
+
+  if (data.number_of_adjustments > 15) {
+    // sometimes sensors fail to change sensitivity. this prevents us from infinite loop
+    ESP_LOGW(TAG, "Too many sensitivity adjustments done. Apparently, sensor reconfiguration fails. Stopping.");
+    return false;
+  }
+  data.number_of_adjustments++;
+
+  // Recommended thresholds as per datasheet
+  static const uint16_t LOW_INTENSITY_THRESHOLD = 1000;
+  static const uint16_t HIGH_INTENSITY_THRESHOLD = 30000;
+  static const AlsGain GAINS[GAINS_COUNT] = {GAIN_1, GAIN_2, GAIN_4, GAIN_8, GAIN_48, GAIN_96};
+  static const IntegrationTime INT_TIMES[TIMES_COUNT] = {
+      INTEGRATION_TIME_50MS,  INTEGRATION_TIME_100MS, INTEGRATION_TIME_150MS, INTEGRATION_TIME_200MS,
+      INTEGRATION_TIME_250MS, INTEGRATION_TIME_300MS, INTEGRATION_TIME_350MS, INTEGRATION_TIME_400MS};
+
+  if (data.ch0 <= LOW_INTENSITY_THRESHOLD) {
+    AlsGain next_gain = get_next(GAINS, data.gain);
+    if (next_gain != data.gain) {
+      data.gain = next_gain;
+      ESP_LOGV(TAG, "Low illuminance. Increasing gain.");
+      return true;
+    }
+    IntegrationTime next_time = get_next(INT_TIMES, data.integration_time);
+    if (next_time != data.integration_time) {
+      data.integration_time = next_time;
+      ESP_LOGV(TAG, "Low illuminance. Increasing integration time.");
+      return true;
+    }
+  } else if (data.ch0 >= HIGH_INTENSITY_THRESHOLD) {
+    AlsGain prev_gain = get_prev(GAINS, data.gain);
+    if (prev_gain != data.gain) {
+      data.gain = prev_gain;
+      ESP_LOGV(TAG, "High illuminance. Decreasing gain.");
+      return true;
+    }
+    IntegrationTime prev_time = get_prev(INT_TIMES, data.integration_time);
+    if (prev_time != data.integration_time) {
+      data.integration_time = prev_time;
+      ESP_LOGV(TAG, "High illuminance. Decreasing integration time.");
+      return true;
+    }
+  } else {
+    ESP_LOGD(TAG, "Illuminance is good enough.");
```

Comment:
Total nitpick, but I think this looks better. haha
```suggestion
    ESP_LOGD(TAG, "Illuminance is sufficient.");
```